### ISSUE TITLE
PR for building for wasm?

### ISSUE THREAD

**Issue by jbr** [rozbb/rust-hpke#21]
Having seen [this blog post](https://blog.cloudflare.com/using-hpke-to-encrypt-request-payloads/) I tried to build this package for wasm with wasm-pack for use from js and discovered that a few changes needed to be made to Cargo.toml in order to make this work. Would you be amenable to a PR with those changes, or alternatively, do you have guidance on how to use this library from js as is?

Thanks!

**Comment by rozbb** [comment]
Yes, I'd be happy to make a feature for wasm compilation! I haven't tried it myself on this crate, so I don't have any insight to share.

**Comment by jbr** [comment]
Thanks!

Did I misunderstand the linked blog post? It read like this crate was designed with wasm in mind. It does seem to build with only small tweaks to Cargo.toml, but I haven't actually used it from js yet.

**Comment by rozbb** [comment]
Interesting, I have no idea what they’re talking about lol. I also wasn’t aware that they did an audit. Good to know they didn’t find anything!

**Comment by migueldemoura** [comment]
> Did I misunderstand the linked blog post? It read like this crate was designed with wasm in mind. It does seem to build with only small tweaks to Cargo.toml, but I haven't actually used it from js yet.

My fault, probably should have made that clearer in the blog post. We did make some small changes to target WASM (as you did in the linked PR) and are indeed successfully using it on our [dashboard](https://dash.cloudflare.com/).

**Comment by jbr** [comment]
@migueldemoura Thanks! That makes sense — I assume you also added a bunch of `#[cfg_attr(feature = "wasm", wasm_bindgen)]`? or something like that? Would it make sense for y'all to pr those changes upstream? Did you also write a js/ts wrapper library or were the autogenerated bindings sufficient for your usecase?

**Comment by migueldemoura** [comment]
Sorry for the delay on this one. I believe that's exactly what we did, but @Margatroid can probably elaborate in more detail since he implemented it.

> Would it make sense for y'all to pr those changes upstream?

We have a small wrapper tool that closely follows one of the [crate's examples](https://github.com/rozbb) as that makes more sense. I believe you came to a [similar conclusion](https://github.com/rozbb/rust-hpke/pull/22#issuecomment-1094360492).

It probably does make sense to have an example of how one would go about using this crate with WASM.
We open-sourced the wrapper tool we use (mostly so people can compile it themselves, it isn't that interesting), but it didn't include the WASM-specific feature since we weren't working on that at the time. I've raised a ticket to get that open-sourced as well.
